### PR TITLE
Compile more YARP node types

### PR DESCRIPTION
**Pull request by jemmaissroff**
Compile more YARP node types including:
- ArrayNode
- AssocSplatNode
- CallNode
- ClassNode
- DefNode
- DefinedNode
- EmbeddedStatementsNode
- FlipFlopNode
- HashNode
- IfNode
- LambdaNode
- LocalVariableReadNode
- LocalVariableWriteNode
- ModuleNode
- MultiWriteNode
- OptionalParameterNode
- RangeNode
- ScopeNode
- SingletonClassNode
- SplatNode
- UnlessNode
- WhileNode
- YieldNode

### CODE REVIEW

File: yarp/yarp_compiler.c [hunk 1]
```diff
@@ -42,6 +59,265 @@ parse_location_symbol(yp_location_t *location) {
...
+    switch(node->type) {
+      case YP_NODE_NIL_NODE:
+        return Qnil;
+      case YP_NODE_TRUE_NODE:
+        return Qtrue;
+      case YP_NODE_FALSE_NODE:
+        return Qfalse;
+        // TODO: Implement this method for the other literal nodes described above
+      default:
+        rb_bug("This node type doesn't have a literal valuej");
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:103]:
```suggestion
        rb_bug("This node type doesn't have a literal value");
```

File: yarp/yarp_compiler.c [hunk 2]
```diff
@@ -97,19 +373,140 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
           yp_begin_node_t *begin_node = (yp_begin_node_t *) node;
           if (begin_node->statements) {
               yp_compile_node(iseq, (yp_node_t *)begin_node->statements, ret, src, popped, compile_context);
           }
           return;
       }
+      case YP_NODE_CALL_NODE: {
+          yp_call_node_t *call_node = (yp_call_node_t *) node;
+
+          ID mid = parse_location_symbol(&call_node->message_loc);
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:426]: We should use `name` here, not `message_loc`. They're slightly different in some cases:

```ruby
foo.()          # name="call" message_loc=nil
foo.bar = 1     # name="bar=" message_loc="bar"
```

File: yarp/yarp_compiler.c [hunk 3]
```diff
@@ -97,19 +373,140 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
+          yp_scope_node_t scope_node;
+          yp_scope_node_init((yp_node_t *)class_node, &scope_node);
+
+          yp_node_t *recursing_node = class_node->constant_path;
+
+          while (recursing_node->type != YP_NODE_CONSTANT_READ_NODE) {
+              recursing_node = ((yp_constant_path_node_t *)recursing_node)->child;
+          }
+
+          ID class_id = parse_location_symbol(&recursing_node->location);
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:484]: Should be able to use the name field on the class for this.

File: yarp/yarp_compiler.c [hunk 4]
```diff
@@ -176,6 +573,40 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
+
+          if (!popped) {
+              ADD_INSN1(ret, &dummy_line_node, putobject, ID2SYM(method_name));
+          }
+          return;
+      }
+      case YP_NODE_DEFINED_NODE: {
+          ADD_INSN(ret, &dummy_line_node, putself);
+          yp_defined_node_t *defined_node = (yp_defined_node_t *)node;
+          enum defined_type dtype = DEFINED_CONST;
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:594]: Could you mark a TODO here?

File: yarp/yarp_compiler.c [hunk 5]
```diff
@@ -312,10 +815,123 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
+          if (!popped) {
+              ADD_GETLOCAL(ret, &dummy_line_node, index, local_read_node->depth);
+          }
+          return;
+      }
+      case YP_NODE_LOCAL_VARIABLE_WRITE_NODE: {
+          yp_local_variable_write_node_t *local_write_node = (yp_local_variable_write_node_t *) node;
+
+          // TODO: Unclear how we get into the case where this has no value
+          if (local_write_node->value) {
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:862]: This always has a `value` now. It was previously incorrectly marked.

File: yarp/yarp_compiler.c [hunk 6]
```diff
@@ -312,10 +815,123 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
+          yp_scope_node_t scope_node;
+          yp_scope_node_init((yp_node_t *)module_node, &scope_node);
+
+          yp_node_t *recursing_node = module_node->constant_path;
+
+          while (recursing_node->type != YP_NODE_CONSTANT_READ_NODE) {
+              recursing_node = ((yp_constant_path_node_t *)recursing_node)->child;
+          }
+
+          ID module_id = parse_location_symbol(&recursing_node->location);
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:900]: Same here with the name field.

**Review by kddnewton — CHANGES_REQUESTED** [review]
A couple small things, otherwise looks good!

File: yarp/yarp_compiler.c (hunk 3 above)
Comment (jemmaissroff) [reply yarp/yarp_compiler.c:484]: Ah yeah nice, I had done this one before we added that!

**Review by kddnewton — APPROVED** [review]

File: yarp/yarp_compiler.c [hunk 7]
```diff
@@ -381,9 +1033,154 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
...
+              optionals_list = parameters_node->optionals;
+          }
+
+          size_t size = locals.size;
+
+          // Index lookup table buffer size is only the number of the locals
+          st_table *index_lookup_table = st_init_numtable();
+
+          VALUE idtmp = 0;
+          rb_ast_id_table_t *tbl = ALLOCV(idtmp, sizeof(rb_ast_id_table_t) + size * sizeof(ID));
```
Comment (kddnewton) [inline yarp/yarp_compiler.c:1056]: This is just a comment for later, but it looks like we _may_ need to free this? Depending on the size. Other places in the codebase use `ALLOCV_END`. We might need to look into this.

**Review by kddnewton — APPROVED** [review]

**Pull request merged**